File: src/com/easy/ConvertSortedArraytoBinarySearchTree_108.java
/**
 * 
 */
package com.easy;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class ConvertSortedArraytoBinarySearchTree_108 {
    // Use binary search concept
    // Time Complexity: O(n)
    // Space Complexity:O(n)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 37.7 MB, less than 54.07%
    public TreeNode sortedArrayToBST(int[] nums) {
        return toBST(nums, 0, nums.length - 1);
    }

    private TreeNode toBST(int[] nums, int mIndexLow, int mIndexHigh) {
        if (mIndexLow > mIndexHigh) {
            return null;
        }
        int mIndexMid = (mIndexLow + mIndexHigh) / 2;
        TreeNode root = new TreeNode(nums[mIndexMid]);
        root.left = toBST(nums, mIndexLow, mIndexMid - 1);
        root.right = toBST(nums, mIndexMid + 1, mIndexHigh);

        return root;
    }
}

File: Golang_src/easy/convertSortedArrayToBinarySearchTree_108.go
package easy

// TreeNode for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Use recursive:
// Time Complexity: O(n)
// Space Complexity: O(n)
// Runtime: 96 ms, faster than 91.25%
// Memory Usage: 233.2 MB, less than 31.31%
func sortedArrayToBST(nums []int) *TreeNode {
	return toBST(nums, 0, len(nums)-1)
}

func toBST(nums []int, start, end int) *TreeNode {
	if start > end {
		return nil
	}
	mid := (start + end) / 2
	node := &TreeNode{Val: nums[mid]}
	node.Left = toBST(nums, start, mid-1)
	node.Right = toBST(nums, mid+1, end)

	return node
}
